streams/src/main/java/org/apache/kafka/streams/processor/internals/StateRestoreCallbackAdapter.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.streams.processor.internals;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.processor.BatchingStateRestoreCallback;
+import org.apache.kafka.streams.processor.StateRestoreCallback;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+public final class StateRestoreCallbackAdapter {
+    private StateRestoreCallbackAdapter() {}
+
+    public static RecordBatchingStateRestoreCallback adapt(final StateRestoreCallback restoreCallback) {
+        if (restoreCallback instanceof RecordBatchingStateRestoreCallback) {
+            return (RecordBatchingStateRestoreCallback) restoreCallback;
+        } else if (restoreCallback instanceof BatchingStateRestoreCallback) {
+            return records -> {
+                final List<KeyValue<byte[], byte[]>> keyValues =
+                    records.stream()
+                           .map(record -> new KeyValue<>(record.key(), record.value()))
+                           .collect(Collectors.toList());
+                ((BatchingStateRestoreCallback) restoreCallback).restoreAll(keyValues);
+            };
+        } else {
+            return records -> {
+                for (final ConsumerRecord<byte[], byte[]> record : records) {
+                    restoreCallback.restore(record.key(), record.value());
```

Comment:
If `restoreCallback` is null we will throw NPE here right?

Reply:
We would throw the NPE later on when `restoreRecords` is actually called, which is what currently happens with the `WrappedBatchingStateRestoreCallback`.

Would you feel better to surface the NPE immediately with a check at the beginning of `adapt`?

Reply:
I think this would be better, unless there's some semantics around registering a null callback.

Reply:
Yeah I felt it is better to guard against null beforehand. This actually come from the unit test where we do `adapt(null).restore(null)` where the `restore` function is supposed to throw not-supported exception, while I was a bit frowning upon it.

Reply:
ok, it's done now.